lib/loglevel.js

```diff
@@ -47,122 +46,169 @@
         }
     }
 
+    // these private functions always need `this` to be set properly
+    
     function enableLoggingWhenConsoleArrives(methodName, level) {
         return function () {
             if (typeof console !== undefinedType) {
-                replaceLoggingMethods(level);
-                self[methodName].apply(self, arguments);
+                replaceLoggingMethods.call(this, level);
+                this[methodName].apply(this, arguments);
             }
         };
     }
 
-    var logMethods = [
-        "trace",
-        "debug",
-        "info",
-        "warn",
-        "error"
-    ];
-
     function replaceLoggingMethods(level) {
         for (var i = 0; i < logMethods.length; i++) {
             var methodName = logMethods[i];
-            self[methodName] = (i < level) ? noop : self.methodFactory(methodName, level);
+            this[methodName] = (i < level) ? noop : this.methodFactory(methodName, level);
         }
     }
 
-    function persistLevelIfPossible(levelNum) {
-        var levelName = (logMethods[levelNum] || 'silent').toUpperCase();
-
-        // Use localStorage if available
-        try {
-            window.localStorage['loglevel'] = levelName;
-            return;
-        } catch (ignore) {}
-
-        // Use session cookie as fallback
-        try {
-            window.document.cookie = "loglevel=" + levelName + ";";
-        } catch (ignore) {}
+    function defaultMethodFactory(methodName, level) {
+        return realMethod(methodName) ||
+               enableLoggingWhenConsoleArrives.call(this, methodName, level);
     }
 
-    function getPersistedLevel() {
-        var storedLevel;
-
-        try {
-            storedLevel = window.localStorage['loglevel'];
-        } catch (ignore) {}
-
-        if (typeof storedLevel === undefinedType) {
-            try {
-                storedLevel = /loglevel=([^;]+)/.exec(window.document.cookie)[1];
-            } catch (ignore) {}
-        }
-
-        // If the stored level is not valid, treat it as if nothing was stored.
-        if (self.levels[storedLevel] === undefined) {
-            storedLevel = undefined;
-        }
+    var logMethods = [
+        "trace",
+        "debug",
+        "info",
+        "warn",
+        "error"
+    ];
 
-        return storedLevel;
+    var defaultLoggerLevel;
+
+    function Logger(name, level, factory) {
```

Comment:
I'd call this 'defaultLevel', since that's what it is. It's not necessarily the actual level of the logger, as it gets overridden by the persisted level if one exists.